src/Constraint/Bound.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver\Constraint;
+
+class Bound
+{
+    /**
+     * @var string
+     */
+    private $version;
+
+    /**
+     * @var bool
+     */
+    private $isInclusive;
+
+    /**
+     * @param string $version
+     * @param bool   $isInclusive
+     */
+    public function __construct($version, $isInclusive)
+    {
+        $this->version = $version;
+        $this->isInclusive = $isInclusive;
+    }
+
+    /**
+     * @return string
+     */
+    public function getVersion()
+    {
+        return $this->version;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isInclusive()
+    {
+        return $this->isInclusive;
+    }
+
+    public function isZero()
+    {
+        return $this->getVersion() === '0' && $this->isInclusive();
+    }
+
+    public function isPositiveInfinity()
+    {
+        return $this->getVersion() === (string) PHP_INT_MAX && !$this->isInclusive();
+    }
+
+    /**
+     * Compares a bound to another with a given operator.
+     *
+     * @param Bound  $other
+     * @param string $operator
+     *
+     * @return bool
+     */
+    public function compareTo(Bound $other, $operator)
+    {
+        if (!in_array($operator, array('<', '>'))) {
+            throw new \InvalidArgumentException('Does not support any other operator other than > or <.');
+        }
+
+        // If they are the same it doesn't matter
+        if ($this == $other) {
+            return false;
+        }
+
+        $compareResult = version_compare($this->getVersion(), $other->getVersion());
+
+        // Not the same version means we don't need to check if the bounds are inclusive or not
+        if (0 !== $compareResult) {
+            return (('>' === $operator) ? 1 : -1) === $compareResult;
+        }
+
+        // Question we're answering here is "am I higher than $other?"
+        return '>' === $operator ? $other->isInclusive() : !$other->isInclusive();
```

Comment:
this looks weird to be that `$this->isInclusive()` is never taken into account. Is this correct ?

Reply:
tbh I don't know. I've been messing with this code for too long but there are lots of tests in the `MultiConstraintTest` which uses the `compareTo()` method internally. So maybe we were missing a constellation that requires a test? Would be great if you could provide that. More eyes is always better :)

Reply:
`$this->isInclusive` is implicitly checked when we do `($this == $other)` above.

When we come to this part of the code, the version on both bounds is the same, so the following table is the result (x means, the bounds are identical, so the early return at the top of the function prevents this from mattering):

|expression|this.isInclusive|other.isInclusive|result|
|-|-|-|-|
|this.compareTo(other, >)|0|0|x|
|this.compareTo(other, >)|0|*1*|*1*|
|this.compareTo(other, >)|1|*0*|*0*|
|this.compareTo(other, >)|1|1|x|
|this.compareTo(other, <)|0|0|x|
|this.compareTo(other, <)|0|*1*|*0*|
|this.compareTo(other, <)|1|*0*|*1*|
|this.compareTo(other, <)|1|1|x|

So you can see that the expression is correct. But since we exclude a large part of the variations with early returns, an expression like `'>'  === $operator ? !$this->isInclusive() : $this->isInclusive()` would be just as correct, since we already made sure that `$this->isInclusive() != $other->isInclusive()` once it gets to this section of the code.